File: 3504.longest-palindrome-after-substring-concatenation-ii.cpp
#
# @lc app=leetcode id=3504 lang=cpp
#
# [3504] Longest Palindrome After Substring Concatenation II
#

# @lc code=start
#include <vector>
#include <string>
#include <algorithm>
#include <utility>
using namespace std;

class Solution {
public:
    int longestPalindrome(string s, string t) {
        int n = s.size();
        int m = t.size();
        int ans = 0;

        // Pal DP for s: ispal and max_start_s
        vector<vector<bool>> ispal_s(n, vector<bool>(n, false));
        vector<int> max_start_s(n, 0);
        for (int i = 0; i < n; ++i) {
            ispal_s[i][i] = true;
            max_start_s[i] = 1;
            ans = max(ans, 1);
        }
        for (int i = 0; i < n - 1; ++i) {
            if (s[i] == s[i + 1]) {
                ispal_s[i][i + 1] = true;
                max_start_s[i] = 2;
                ans = max(ans, 2);
            }
        }
        for (int leng = 3; leng <= n; ++leng) {
            for (int i = 0; i <= n - leng; ++i) {
                int j = i + leng - 1;
                if (s[i] == s[j] && ispal_s[i + 1][j - 1]) {
                    ispal_s[i][j] = true;
                    max_start_s[i] = max(max_start_s[i], leng);
                    ans = max(ans, leng);
                }
            }
        }

        // Pal DP for t: ispal, max_start_t (unused), max_end_t
        vector<vector<bool>> ispal_t(m, vector<bool>(m, false));
        vector<int> max_start_t(m, 0);
        vector<int> max_end_t(m, 0);
        for (int i = 0; i < m; ++i) {
            ispal_t[i][i] = true;
            max_start_t[i] = 1;
            max_end_t[i] = 1;
            ans = max(ans, 1);
        }
        for (int i = 0; i < m - 1; ++i) {
            if (t[i] == t[i + 1]) {
                ispal_t[i][i + 1] = true;
                max_start_t[i] = 2;
                max_end_t[i + 1] = 2;
                ans = max(ans, 2);
            }
        }
        for (int leng = 3; leng <= m; ++leng) {
            for (int i = 0; i <= m - leng; ++i) {
                int j = i + leng - 1;
                if (t[i] == t[j] && ispal_t[i + 1][j - 1]) {
                    ispal_t[i][j] = true;
                    max_start_t[i] = max(max_start_t[i], leng);
                    max_end_t[j] = max(max_end_t[j], leng);
                    ans = max(ans, leng);
                }
            }
        }

        // Type1: match1[i][j] longest s[i...] prefix == rev(t[...j] suffix)
        vector<vector<int>> match1(n, vector<int>(m, 0));
        for (int i = n - 1; i >= 0; --i) {
            for (int j = 0; j < m; ++j) {
                if (s[i] == t[j]) {
                    int nextk = 0;
                    if (i + 1 < n && j - 1 >= 0) {
                        nextk = match1[i + 1][j - 1];
                    }
                    match1[i][j] = 1 + nextk;
                }
            }
        }
        // Process type1
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                int k = match1[i][j];
                if (k > 0) {
                    int mid_start = i + k;
                    int mid_len = 0;
                    if (mid_start < n) {
                        mid_len = max_start_s[mid_start];
                    }
                    ans = max(ans, 2 * k + mid_len);
                }
            }
        }

        // Hashing for type2
        const long long MOD1 = 1000000007LL;
        const long long BASE1 = 131LL;
        const long long MOD2 = 1000000009LL;
        const long long BASE2 = 137LL;

        auto make_hasher = [&](const string& str) -> auto {
            int sz = str.size();
            vector<long long> p1(sz + 1, 1), h1(sz + 1, 0);
            vector<long long> p2(sz + 1, 1), h2(sz + 1, 0);
            for (int i = 1; i <= sz; ++i) {
                p1[i] = p1[i - 1] * BASE1 % MOD1;
                h1[i] = (h1[i - 1] * BASE1 + (str[i - 1] - 'a' + 1)) % MOD1;
                p2[i] = p2[i - 1] * BASE2 % MOD2;
                h2[i] = (h2[i - 1] * BASE2 + (str[i - 1] - 'a' + 1)) % MOD2;
            }
            auto get = [&](int L, int R) -> pair<long long, long long> {
                int len = R - L + 1;
                long long val1 = (h1[R + 1] + MOD1 - h1[L] * p1[len] % MOD1) % MOD1;
                long long val2 = (h2[R + 1] + MOD2 - h2[L] * p2[len] % MOD2) % MOD2;
                return {val1, val2};
            };
            return get;
        };

        string tr = t;
        reverse(tr.begin(), tr.end());
        auto get_hs = make_hasher(s);
        auto get_htr = make_hasher(tr);

        // Helper to get max k for type2 at p, r
        auto get_k = [&](int p, int r) -> int {
            int low = 1;
            int hi = min(p + 1, r + 1);
            int res = 0;
            while (low <= hi) {
                int md = low + (hi - low) / 2;
                int starts = p - md + 1;
                if (starts < 0) {
                    hi = md - 1;
                    continue;
                }
                int starttr = m - 1 - r;
                int endtr = starttr + md - 1;
                if (endtr >= m) {
                    hi = md - 1;
                    continue;
                }
                auto hs_val = get_hs(starts, p);
                auto htr_val = get_htr(starttr, endtr);
                if (hs_val == htr_val) {
                    res = md;
                    low = md + 1;
                } else {
                    hi = md - 1;
                }
            }
            return res;
        };

        // Process type2
        for (int p = 0; p < n; ++p) {
            for (int r = 0; r < m; ++r) {
                int k = get_k(p, r);
                if (k > 0) {
                    int mid_end = r - k;
                    int mid_len = 0;
                    if (mid_end >= 0) {
                        mid_len = max_end_t[mid_end];
                    }
                    ans = max(ans, 2 * k + mid_len);
                }
            }
        }

        return ans;
    }
};
# @lc code=end

File: 3504.longest-palindrome-after-substring-concatenation-ii.py
#
# @lc app=leetcode id=3504 lang=python3
#
# [3504] Longest Palindrome After Substring Concatenation II
#

# @lc code=start
class Solution:
    def longestPalindrome(self, s: str, t: str) -> int:
        def get_prefix(string: str, base: int, mod: int) -> tuple[list[int], list[int]]:
            n = len(string)
            prefix = [0] * (n + 1)
            power = [1] * (n + 1)
            for i in range(n):
                prefix[i + 1] = (prefix[i] * base + (ord(string[i]) - ord('a') + 1)) % mod
                power[i + 1] = (power[i] * base) % mod
            return prefix, power

        def get_hash(prefix: list[int], power: list[int], mod: int, left: int, right: int) -> int:
            res = (prefix[right] - prefix[left] * power[right - left] % mod + mod) % mod
            return res

        def compute_max_and_forward(string: str) -> tuple[int, list[int]]:
            n = len(string)
            if n == 0:
                return 0, []
            dp = [[False] * n for _ in range(n)]
            forward = [1] * n
            max_len = 1
            for i in range(n):
                dp[i][i] = True
            for i in range(n - 1):
                if string[i] == string[i + 1]:
                    dp[i][i + 1] = True
                    forward[i] = 2
                    max_len = 2
            for length in range(3, n + 1):
                for i in range(n - length + 1):
                    j = i + length - 1
                    if string[i] == string[j] and dp[i + 1][j - 1]:
                        dp[i][j] = True
                        forward[i] = max(forward[i], length)
                        max_len = max(max_len, length)
            return max_len, forward

        def compute_max_and_backward(string: str) -> tuple[int, list[int]]:
            n = len(string)
            if n == 0:
                return 0, []
            dp = [[False] * n for _ in range(n)]
            backward = [1] * n
            max_len = 1
            for i in range(n):
                dp[i][i] = True
            for i in range(n - 1):
                if string[i] == string[i + 1]:
                    dp[i][i + 1] = True
                    backward[i + 1] = 2
                    max_len = 2
            for length in range(3, n + 1):
                for i in range(n - length + 1):
                    j = i + length - 1
                    if string[i] == string[j] and dp[i + 1][j - 1]:
                        dp[i][j] = True
                        backward[j] = max(backward[j], length)
                        max_len = max(max_len, length)
            return max_len, backward

        ns = len(s)
        nt = len(t)
        max_s, forward_s = compute_max_and_forward(s)
        max_t, backward_t = compute_max_and_backward(t)
        ans = max(max_s, max_t)

        MOD1 = 10**9 + 7
        MOD2 = 10**9 + 9
        BASE1 = 31
        BASE2 = 37

        prefix_s1, power_s1 = get_prefix(s, BASE1, MOD1)
        prefix_s2, power_s2 = get_prefix(s, BASE2, MOD2)
        tr = t[::-1]
        prefix_tr1, power_tr1 = get_prefix(tr, BASE1, MOD1)
        prefix_tr2, power_tr2 = get_prefix(tr, BASE2, MOD2)

        def get_h_s(left: int, right: int, which: int) -> int:
            if which == 1:
                return get_hash(prefix_s1, power_s1, MOD1, left, right)
            return get_hash(prefix_s2, power_s2, MOD2, left, right)

        def get_h_tr(left: int, right: int, which: int) -> int:
            if which == 1:
                return get_hash(prefix_tr1, power_tr1, MOD1, left, right)
            return get_hash(prefix_tr2, power_tr2, MOD2, left, right)

        max_k = min(ns, nt)
        for k in range(1, max_k + 1):
            # Case 1: longer or equal in s (middle in s)
            set_tr_k = set()
            for j in range(nt - k + 1):
                h1 = get_h_tr(j, j + k, 1)
                h2 = get_h_tr(j, j + k, 2)
                set_tr_k.add((h1, h2))
            for start in range(ns - k + 1):
                hs1 = get_h_s(start, start + k, 1)
                hs2 = get_h_s(start, start + k, 2)
                if (hs1, hs2) in set_tr_k:
                    pos = start + k
                    m = forward_s[pos] if pos < ns else 0
                    ans = max(ans, 2 * k + m)

            # Case 2: longer or equal in t (middle in t)
            set_s_k = set()
            for b in range(ns - k + 1):
                h1 = get_h_s(b, b + k, 1)
                h2 = get_h_s(b, b + k, 2)
                set_s_k.add((h1, h2))
            for suf_start in range(nt - k + 1):
                j_start = nt - suf_start - k
                j_end = nt - suf_start
                htr1 = get_h_tr(j_start, j_end, 1)
                htr2 = get_h_tr(j_start, j_end, 2)
                if (htr1, htr2) in set_s_k:
                    p_end = suf_start - 1
                    p = backward_t[p_end] if p_end >= 0 else 0
                    ans = max(ans, 2 * k + p)
        return ans

# @lc code=end